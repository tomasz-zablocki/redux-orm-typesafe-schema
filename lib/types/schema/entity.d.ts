import { Class } from 'utility-types';
import { AttributeField, EntityBuilders, Serializable, Supplier } from './fields';
import { Repository, Session } from '../redux-orm';
import { AnyAction } from 'redux';
import OneToOneBuilder = EntityBuilders.OneToOneBuilder;
import ManyToManyBuilder = EntityBuilders.ManyToManyBuilder;
import ManyToOneBuilder = EntityBuilders.ManyToOneBuilder;
import OneToManyBuilder = EntityBuilders.OneToManyBuilder;
export { FieldType, Field, RelationType, RelationField, AttributeField, OneToOneField, OneToManyField, ManyToOneField, ManyToManyField, EntityKeys, EntityBuilders } from './fields';
declare abstract class Entity<E extends Entity<E>> {
    abstract readonly modelName: string;
    entityClass(): Class<E>;
    reduce(action: AnyAction, repository: Repository<E>, repositories?: Session<any>): void;
    attribute<ValueType extends Serializable>(supplier?: Supplier<ValueType>): AttributeField<E, ValueType>;
    oneToOne: OneToOneBuilder<E>;
    oneToMany: OneToManyBuilder<E>;
    manyToOne: ManyToOneBuilder<E>;
    manyToMany: ManyToManyBuilder<E>;
}
export { Entity };
