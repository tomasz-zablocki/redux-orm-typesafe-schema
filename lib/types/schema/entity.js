"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Entity = (function () {
    function Entity() {
        var _this = this;
        this.oneToOne = function (to) {
            return {
                ref: function (reverseField) { return ({
                    fieldType: 'OneToOne',
                    virtual: false,
                    from: _this.entityClass(),
                    to: to,
                    reverseField: reverseField
                }); },
                virtualRef: function (reverseField) { return ({
                    fieldType: 'OneToOne',
                    virtual: true,
                    from: _this.entityClass(),
                    to: to,
                    reverseField: reverseField
                }); }
            };
        };
        this.oneToMany = function (to) { return ({
            ref: function (reverseField) { return ({
                fieldType: 'OneToMany',
                virtual: true,
                from: _this.entityClass(),
                to: to,
                reverseField: reverseField
            }); }
        }); };
        this.manyToOne = function (to) {
            var build = function (options) { return (tslib_1.__assign({ fieldType: 'ManyToOne', virtual: false, from: _this.entityClass(), to: to }, options)); };
            return {
                ref: function (reverseField) { return build({ reverseField: reverseField }); },
                noRef: build
            };
        };
        this.manyToMany = function (to) { return ({
            through: function (through) { return ({
                virtualRef: function (reverseField, throughReverseField) { return ({
                    fieldType: 'ManyToMany',
                    virtual: true,
                    from: _this.entityClass(),
                    to: to,
                    through: through,
                    reverseField: reverseField,
                    throughReverseField: throughReverseField
                }); },
                ref: function (reverseField, throughReverseField) { return ({
                    fieldType: 'ManyToMany',
                    virtual: false,
                    from: _this.entityClass(),
                    to: to,
                    through: through,
                    reverseField: reverseField,
                    throughReverseField: throughReverseField
                }); }
            }); }
        }); };
    }
    Entity.prototype.entityClass = function () {
        return this.constructor;
    };
    Entity.prototype.reduce = function (action, repository, repositories) { };
    Entity.prototype.attribute = function (supplier) {
        return {
            fieldType: 'Attribute',
            from: this.entityClass(),
            virtual: false,
            supplier: supplier
        };
    };
    return Entity;
}());
exports.Entity = Entity;
//# sourceMappingURL=entity.js.map