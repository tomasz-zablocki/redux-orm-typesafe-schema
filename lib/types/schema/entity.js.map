{"version":3,"file":"entity.js","sourceRoot":"","sources":["../../../src/types/schema/entity.ts"],"names":[],"mappings":";;;AAuBA;IAAA;QAAA,iBA0FC;QAlEC,aAAQ,GAAuB,UAAA,EAAE;YAC/B,OAAO;gBACL,GAAG,EAAE,UAAA,YAAY,IAAI,OAAA,CAAC;oBACpB,SAAS,EAAE,UAAU;oBACrB,OAAO,EAAE,KAAK;oBACd,IAAI,EAAE,KAAI,CAAC,WAAW,EAAE;oBACxB,EAAE,IAAA;oBACF,YAAY,cAAA;iBACb,CAAC,EANmB,CAMnB;gBACF,UAAU,EAAE,UAAA,YAAY,IAAI,OAAA,CAAC;oBAC3B,SAAS,EAAE,UAAU;oBACrB,OAAO,EAAE,IAAI;oBACb,IAAI,EAAE,KAAI,CAAC,WAAW,EAAE;oBACxB,EAAE,IAAA;oBACF,YAAY,cAAA;iBACb,CAAC,EAN0B,CAM1B;aACH,CAAA;QACH,CAAC,CAAA;QAED,cAAS,GAAwB,UAAA,EAAE,IAAI,OAAA,CAAC;YACtC,GAAG,EAAE,UAAA,YAAY,IAAI,OAAA,CAAC;gBACpB,SAAS,EAAE,WAAW;gBACtB,OAAO,EAAE,IAAI;gBACb,IAAI,EAAE,KAAI,CAAC,WAAW,EAAE;gBACxB,EAAE,IAAA;gBACF,YAAY,cAAA;aACb,CAAC,EANmB,CAMnB;SACH,CAAC,EARqC,CAQrC,CAAA;QAEF,cAAS,GAAwB,UAAA,EAAE;YACjC,IAAM,KAAK,GAAG,UAAC,OAAY,IAAK,OAAA,oBAC9B,SAAS,EAAE,WAAoB,EAC/B,OAAO,EAAE,KAAK,EACd,IAAI,EAAE,KAAI,CAAC,WAAW,EAAE,EACxB,EAAE,IAAA,IACC,OAAO,EACV,EAN8B,CAM9B,CAAA;YAEF,OAAO;gBACL,GAAG,EAAE,UAAA,YAAY,IAAI,OAAA,KAAK,CAAC,EAAE,YAAY,cAAA,EAAE,CAAC,EAAvB,CAAuB;gBAC5C,KAAK,EAAE,KAAK;aACb,CAAA;QACH,CAAC,CAAA;QAED,eAAU,GAAyB,UAAA,EAAE,IAAI,OAAA,CAAC;YACxC,OAAO,EAAE,UAAA,OAAO,IAAI,OAAA,CAAC;gBACnB,UAAU,EAAE,UAAC,YAAY,EAAE,mBAAmB,IAAK,OAAA,CAAC;oBAClD,SAAS,EAAE,YAAY;oBACvB,OAAO,EAAE,IAAI;oBACb,IAAI,EAAE,KAAI,CAAC,WAAW,EAAE;oBACxB,EAAE,IAAA;oBACF,OAAO,SAAA;oBACP,YAAY,cAAA;oBACZ,mBAAmB,qBAAA;iBACpB,CAAC,EARiD,CAQjD;gBACF,GAAG,EAAE,UAAC,YAAY,EAAE,mBAAmB,IAAK,OAAA,CAAC;oBAC3C,SAAS,EAAE,YAAY;oBACvB,OAAO,EAAE,KAAK;oBACd,IAAI,EAAE,KAAI,CAAC,WAAW,EAAE;oBACxB,EAAE,IAAA;oBACF,OAAO,SAAA;oBACP,YAAY,cAAA;oBACZ,mBAAmB,qBAAA;iBACpB,CAAC,EAR0C,CAQ1C;aACH,CAAC,EAnBkB,CAmBlB;SACH,CAAC,EArBuC,CAqBvC,CAAA;IACJ,CAAC;IAvFC,4BAAW,GAAX;QACE,OAAO,IAAI,CAAC,WAAuB,CAAA;IACrC,CAAC;IAED,uBAAM,GAAN,UACE,MAAiB,EACjB,UAAyB,EACzB,YAA2B,IACpB,CAAC;IAEV,0BAAS,GAAT,UACE,QAA8B;QAE9B,OAAO;YACL,SAAS,EAAE,WAAW;YACtB,IAAI,EAAE,IAAI,CAAC,WAAW,EAAE;YACxB,OAAO,EAAE,KAAK;YACd,QAAQ,UAAA;SACT,CAAA;IACH,CAAC;IAoEH,aAAC;AAAD,CAAC,AA1FD,IA0FC;AAEQ,wBAAM"}