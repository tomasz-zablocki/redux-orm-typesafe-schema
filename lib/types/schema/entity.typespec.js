"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("@spec/utils");
describe('Entity class ->', function () {
    it('Vod ->', function () {
        utils_1.testType();
        utils_1.testType();
        utils_1.testType();
        utils_1.testType();
        utils_1.testType();
        utils_1.testType();
        utils_1.testType();
        utils_1.testType();
        utils_1.testType();
    });
});
//# sourceMappingURL=entity.typespec.js.map