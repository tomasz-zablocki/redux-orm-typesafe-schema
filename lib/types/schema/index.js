"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var entity_1 = require("./entity");
exports.Entity = entity_1.Entity;
//# sourceMappingURL=index.js.map