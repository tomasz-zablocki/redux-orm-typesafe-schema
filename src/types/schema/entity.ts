import { Class } from 'utility-types'
import {
  AttributeField,
  EntityBuilders,
  EntityKeys,
  RelationType,
  ReverseRelation,
  Serializable,
  Supplier
} from './fields'
import { Repository, Session } from '../redux-orm'
import { AnyAction } from 'redux'
import RelationBuilder = EntityBuilders.RelationBuilder
import ManyToManyBuilder = EntityBuilders.ManyToManyBuilder

export {
  FieldType,
  Field,
  RelationType,
  RelationField,
  AttributeField,
  ManyToManyField,
  EntityKeys,
  EntityBuilders
} from './fields'

abstract class Entity<E extends Entity<E>> {
  abstract readonly modelName: string

  entityClass(): Class<E> {
    return this.constructor as Class<E>
  }

  reduce(
    action: AnyAction,
    repository: Repository<E>,
    repositories?: Session<any>
  ): void {}

  attribute<ValueType extends Serializable>(
    supplier?: Supplier<ValueType>
  ): AttributeField<E, ValueType> {
    return {
      fieldType: 'Attribute',
      from: this.entityClass(),
      supplier
    }
  }

  oneToOne: RelationBuilder<E, 'OneToOne'> = to => ({
    ref: reverseField => this._ref(to, 'OneToOne', reverseField)
  })

  oneToMany: RelationBuilder<E, 'OneToMany'> = to => ({
    ref: reverseField => this._ref(to, 'OneToMany', reverseField)
  })

  manyToOne: RelationBuilder<E, 'ManyToOne'> = to => ({
    ref: reverseField => this._ref(to, 'ManyToOne', reverseField)
  })

  manyToMany: ManyToManyBuilder<E> = to => ({
    through: through => ({
      ref: (reverseField, throughReverseField) => ({
        ...this._ref(to, 'ManyToMany', reverseField),
        through,
        throughReverseField
      })
    })
  })

  private _ref = <To extends Entity<To>, TRelation extends RelationType>(
    to: Class<To>,
    fieldType: TRelation,
    reverseField?: EntityKeys.RelationKeys<To, E, ReverseRelation<TRelation>>
  ) => ({ to, from: this.entityClass(), reverseField, fieldType })
}

export { Entity }
