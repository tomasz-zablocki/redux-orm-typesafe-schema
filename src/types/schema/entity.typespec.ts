// noinspection ES6UnusedImports
import { $Keys, Class, Unionize } from 'utility-types'
// noinspection ES6UnusedImports
import {
  AttributeField,
  EntityBuilders,
  EntityKeys,
  Field,
  FieldType,
  ManyToManyField,
  ManyToOneField,
  OneToManyField,
  OneToOneField,
  RelationField,
  RelationType
} from './fields'
// noinspection ES6UnusedImports
import {
  Orm,
  OrmModel,
  OrmModelClass,
  OrmSession,
  OrmState
} from '../redux-orm/aliases'
// noinspection ES6UnusedImports
import { Content, schema, Source, Vod, VodId } from '@spec/schema.fixture'
import { testType } from '@spec/utils'
import { Entity } from './entity'

// @dts-jest:group Types
describe('Entity class spec', () => {
  it('Entity ', () => {
    // @dts-jest:pass:snap
    testType<Vod>()
    // @dts-jest:pass:snap
    testType<Entity<Vod>>()
    // @dts-jest:pass:snap
    testType<Class<Vod>>()

    const vodClass = new Vod().entityClass()
    // @dts-jest:pass:snap
    testType<typeof vodClass>()
    // @dts-jest:pass:snap
    testType<Entity<Vod>>()

    // @dts-jest:pass:snap
    testType<Vod['sources']>()
    // @dts-jest:pass:snap
    testType<Vod['title']>()
    // @dts-jest:pass:snap
    testType<Vod['links']>()
    // @dts-jest:pass:snap
    testType<ReturnType<Required<Vod['links']>['supplier']>>()

    // @dts-jest:pass:snap
    testType<Vod['links']>()
  })
})
