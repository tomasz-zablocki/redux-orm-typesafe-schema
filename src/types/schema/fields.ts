import { Entity } from './entity'
import { Class, PickByValue, SetDifference } from 'utility-types'

type Primitive = number | string | boolean

type Serializable =
  | Primitive
  | Primitive[]
  | undefined
  | {
      [K: string]: Serializable | Serializable[]
    }
type Supplier<T> = () => T

export {
  Primitive,
  Serializable,
  FieldType,
  Field,
  Supplier,
  AttributeField,
  RelationType,
  RelationField,
  ManyToManyField,
  EntityKeys,
  EntityBuilders
}

type FieldType =
  | 'Attribute'
  | 'OneToOne'
  | 'OneToMany'
  | 'ManyToOne'
  | 'ManyToMany'

type Field =
  | AttributeField<any, any>
  | RelationField<any, any, any, any>
  | ManyToManyField<any, any, any, any, any>

interface EntityField<FType extends FieldType, From extends Entity<From>> {
  readonly fieldType: FType
  readonly virtual: boolean
  readonly from: Class<From>
}

interface AttributeField<
  From extends Entity<From>,
  ValueType extends Serializable
> extends EntityField<'Attribute', From> {
  readonly fieldType: 'Attribute'
  readonly supplier?: Supplier<ValueType>
  readonly virtual: false
}

type RelationType = SetDifference<FieldType, 'Attribute'>

interface RelationField<
  TRelationType extends RelationType,
  From extends Entity<From>,
  To extends Entity<To>,
  TReverseField extends keyof To
> extends EntityField<TRelationType, From> {
  readonly to: Class<To>
  readonly reverseField?: TReverseField
}

namespace EntityKeys {
  export type RelationKeys<
    From extends Entity<From>,
    To extends Entity<To>,
    TRelationType extends RelationType
  > = {
    [K in keyof From]-?: Required<From>[K] extends RelationField<
      TRelationType,
      From,
      To,
      infer R
    >
      ? K
      : never
  }[keyof From]

  export type Attribute<E extends Entity<E>> = keyof PickByValue<
    Required<E>,
    AttributeField<E, any>
  >

  export type Relation<
    E extends Entity<E>,
    TRelationType extends RelationType
  > = keyof PickByValue<Required<E>, RelationField<TRelationType, E, any, any>>
}

interface ManyToManyField<
  From extends Entity<From>,
  To extends Entity<To>,
  Through extends Entity<Through>,
  TReverseField extends EntityKeys.RelationKeys<To, From, 'ManyToMany'>,
  TThroughReverseField extends EntityKeys.RelationKeys<To, From, 'ManyToMany'>
> extends RelationField<'ManyToMany', From, To, TReverseField> {
  readonly fieldType: 'ManyToMany'
  readonly through: Class<Through>
  readonly throughReverseField: TThroughReverseField
}

namespace EntityBuilders {
  export interface OneToOneBuilder<From extends Entity<From>> {
    <To extends Entity<To>>(to: Class<To>): {
      ref(
        reverseField: EntityKeys.RelationKeys<To, From, 'OneToOne'>
      ): RelationField<'OneToOne', From, To, any>
      virtualRef(
        reverseField: EntityKeys.RelationKeys<To, From, 'OneToOne'>
      ): RelationField<'OneToOne', From, To, any>
    }
  }

  export interface OneToManyBuilder<From extends Entity<From>> {
    <To extends Entity<To>>(to: Class<To>): {
      ref(
        reverseField: EntityKeys.RelationKeys<To, From, 'ManyToOne'>
      ): RelationField<'OneToMany', From, To, any>
    }
  }

  export interface ManyToOneBuilder<From extends Entity<From>> {
    <To extends Entity<To>>(to: Class<To>): {
      noRef(): RelationField<'ManyToOne', From, To, any>
      ref(
        reverseField: EntityKeys.RelationKeys<To, From, 'OneToMany'>
      ): RelationField<'ManyToOne', From, To, any>
    }
  }

  export interface ManyToManyBuilder<From extends Entity<From>> {
    <To extends Entity<To>>(to: Class<To>): {
      through: <Through extends Entity<Through>>(
        through: Class<Through>
      ) => {
        ref(
          reverseField: EntityKeys.RelationKeys<To, From, 'ManyToMany'>,
          throughField: EntityKeys.RelationKeys<Through, From, 'ManyToOne'>
        ): ManyToManyField<From, To, Through, any, any>
        virtualRef(
          reverseField: EntityKeys.RelationKeys<To, From, 'ManyToMany'>,
          throughField: EntityKeys.RelationKeys<Through, From, 'ManyToOne'>
        ): ManyToManyField<From, To, Through, any, any>
      }
    }
  }
}
