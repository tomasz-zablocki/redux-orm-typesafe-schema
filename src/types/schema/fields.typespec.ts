import { Book, ISBN, Person } from '@spec/schema.fixture'
import { testType } from '@spec/utils'
import { EntityKeys, RelationType } from './fields'

// @dts-jest:group EntityKeys ->
describe('fields', () => {
  it('EntityKeys ->', () => {
    // @dts-jest:pass:snap Attribute<Book>
    testType<EntityKeys.Attribute<Book>>()

    // @dts-jest:pass:snap Relation<Book, *>
    testType<EntityKeys.Relation<Book, RelationType>>()

    // @dts-jest:pass:snap Relation<Book, OneToOne>
    testType<EntityKeys.Relation<Book, 'OneToOne'>>()

    // @dts-jest:pass:snap Relation<Book, OneToMany>
    testType<EntityKeys.Relation<Book, 'OneToMany'>>()

    // @dts-jest:pass:snap Relation<Book, ManyToOne>
    testType<EntityKeys.Relation<Book, 'ManyToOne'>>()

    // @dts-jest:pass:snap Relation<Book, ManyToMany>
    testType<EntityKeys.Relation<Book, 'ManyToMany'>>()

    // @dts-jest:pass:snap OneToOne<Book, ISBN>
    testType<EntityKeys.RelationKeys<Book, ISBN, 'OneToOne'>>()

    // @dts-jest:pass:snap ManyToMany<Book, Person>
    testType<EntityKeys.RelationKeys<Book, Person, 'ManyToMany'>>()

    // @dts-jest:pass:snap ManyToMany<Person, Book>
    testType<EntityKeys.RelationKeys<Person, Book, 'ManyToMany'>>()
  })
})
