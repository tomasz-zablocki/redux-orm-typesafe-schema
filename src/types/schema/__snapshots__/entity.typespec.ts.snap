// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity class -> testType<Book>() (type) should match snapshot 1`] = `"Book"`;

exports[`Entity class -> testType<Book['authors']>() (type) should match snapshot 1`] = `"ManyToManyField<Book, import(\\"src/spec/schema.fixture\\").Person, import(\\"src/spec/schema.fixture\\").Authorship, any, any>"`;

exports[`Entity class -> testType<Book['genre']>() (type) should match snapshot 1`] = `"RelationField<\\"ManyToOne\\", Book, import(\\"src/spec/schema.fixture\\").Genre, any>"`;

exports[`Entity class -> testType<Book['id']>() (type) should match snapshot 1`] = `"AttributeField<Book, string>"`;

exports[`Entity class -> testType<Book['links']>() (type) should match snapshot 1`] = `"AttributeField<Book, { description?: string | undefined; cover?: string | undefined; }>"`;

exports[`Entity class -> testType<Book['title']>() (type) should match snapshot 1`] = `"AttributeField<Book, string>"`;

exports[`Entity class -> testType<ReturnType<Book['entityClass']>>() (type) should match snapshot 1`] = `"Class<Book>"`;
